### PR TITLE
Widget Upgrade: Compatibility Schema Versions

### PR DESCRIPTION
**Pull request by ifarzana**
- Related to https://github.com/xibosignageltd/xibo-private/issues/286

### CODE REVIEW

**Comment by ifarzana** [comment]
@dasgarner I have pointed a question in TODO, can you check please?

File: lib/XTR/WidgetCompatibilityTask.php [hunk 1]
```diff
@@ -71,6 +71,9 @@ public function run()
         // If the Widget Compatibility class is defined, it needs to be executed to upgrade the widgets.
         $this->runMessage = '# ' . __('Widget Compatibility') . PHP_EOL . PHP_EOL;
 
+        // We should get all widgets which are < the schema version of the module installed, and
+        // upgrade them to the schema version of the module installed.
+        // TODO How to get widgets which are < the schema version of the module installed?
```
Comment (dasgarner) [inline lib/XTR/WidgetCompatibilityTask.php:76]: Did you have an answer to this?

File: lib/XTR/WidgetCompatibilityTask.php [hunk 2]
```diff
@@ -139,7 +109,115 @@ public function run()
...
+                        $playlist->notifyLayouts();
+                    } catch (\Exception $e) {
+                        $this->log->error('Failed to set layout rebuild for widgetId: ' . $widget->widgetId .
+                            ', message: ' . $e->getMessage());
+                        $this->appendRunMessage('Layout rebuild error for widgetId: : '. $widget->widgetId);
+                    }
+
+                    $this->store->commitIfNecessary();
+                }
+
```
Comment (dasgarner) [inline lib/XTR/WidgetCompatibilityTask.php:214]: phpcs ;-)

File: lib/XTR/WidgetCompatibilityTask.php [hunk 3]
```diff
@@ -139,7 +109,115 @@ public function run()
...
+                        // Get the layout of the widget and set it to rebuild.
+                        $playlist = $this->playlistFactory->getById($widget->playlistId);
+                        $playlist->notifyLayouts();
+                    } catch (\Exception $e) {
+                        $this->log->error('Failed to set layout rebuild for widgetId: ' . $widget->widgetId .
+                            ', message: ' . $e->getMessage());
+                        $this->appendRunMessage('Layout rebuild error for widgetId: : '. $widget->widgetId);
+                    }
+
+                    $this->store->commitIfNecessary();
```
Comment (dasgarner) [inline lib/XTR/WidgetCompatibilityTask.php:212]: We're doing this after every widget, but is that strictly necessary?

File: lib/XTR/WidgetCompatibilityTask.php [hunk 4]
```diff
@@ -139,7 +109,115 @@ public function run()
...
+        $sql = '
+          SELECT widget.widgetId,
+              widget.playlistId,
+              widget.ownerId,
+              widget.type,
+              widget.duration,
+              widget.displayOrder,
+              widget.schemaVersion
+          FROM `widget` 
+          WHERE `widget`.type = :type
```
Comment (dasgarner) [inline lib/XTR/WidgetCompatibilityTask.php:132]: For compatibility with MySQL8 we should have backticks around the `type` column

File: lib/XTR/WidgetCompatibilityTask.php [hunk 5]
```diff
@@ -139,7 +109,115 @@ public function run()
...
+    /**
+     *
+     * @param string $type
+     * @param int $version
+     * @return false|\PDOStatement
+     */
+    private function executeStatement(string $type, int $version): bool|\PDOStatement
+    {
+        $sql = '
+          SELECT widget.widgetId,
```
Comment (dasgarner) [inline lib/XTR/WidgetCompatibilityTask.php:124]: I can see you're using `widgetId` - but what are you using the other columns for?

**Review by dasgarner — COMMENTED** [review]
This is moving in the right direction.

Do we need to handle anything in layout import? So that we do not upgrade widgets which are already at the right schema version?

**Review by dasgarner — APPROVED** [review]

**Pull request merged**